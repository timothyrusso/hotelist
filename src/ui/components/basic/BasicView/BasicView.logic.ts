import { Platforms } from "@/src/ui/constants/platform/Platforms";
import { StatusBar } from "react-native";
import { styles } from "./BasicView.style";

export const useBasicViewLogic = (isFullScreen: boolean) => {
	const paddingTop =
		Platforms.Android && !isFullScreen ? StatusBar.currentHeight : 0;

	const basicViewStyle = styles(paddingTop ?? 0);

	return { basicViewStyle };
};
