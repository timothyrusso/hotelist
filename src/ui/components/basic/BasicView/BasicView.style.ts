import { Colors } from "@/src/ui/constants/style/Colors";
import { StyleSheet } from "react-native";

export const styles = (paddingTop: number) =>
	StyleSheet.create({
		containerViewStyle: {
			width: "100%",
			flex: 1,
			justifyContent: "center",
			alignItems: "center",
			backgroundColor: Colors.white.default,
			paddingTop,
		},
	});
