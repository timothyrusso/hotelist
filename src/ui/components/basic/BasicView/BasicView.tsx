import { StatusBar } from "expo-status-bar";
import type { FC, ReactElement } from "react";
import { SafeAreaView, View } from "react-native";
import { useBasicViewLogic } from "./BasicView.logic";

type BasicViewProps = {
	children: ReactElement;
	statusBarStyle?: "light" | "dark";
	isFullScreen?: boolean;
	nameView: string;
};

export const BasicView: FC<BasicViewProps> = ({
	children,
	statusBarStyle = "dark",
	isFullScreen = false,
	nameView,
}) => {
	const { basicViewStyle } = useBasicViewLogic(isFullScreen);
	const Container = isFullScreen ? View : SafeAreaView;

	// Analyze ri-renders
	console.log(`----------${nameView}---------`);

	return (
		<Container style={basicViewStyle.containerViewStyle}>
			<StatusBar style={statusBarStyle} />
			{children}
		</Container>
	);
};
