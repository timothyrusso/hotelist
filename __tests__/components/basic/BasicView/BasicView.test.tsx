import { BasicView } from "@/src/ui/components/basic/BasicView/BasicView";
import { render } from "@testing-library/react-native";
import { StatusBar } from "expo-status-bar";
import { ScrollView, Text, View } from "react-native";

// Mock expo-status-bar to avoid platform-specific issues
jest.mock("expo-status-bar", () => ({
	StatusBar: jest.fn(() => null),
}));

describe("BasicView", () => {
	beforeEach(() => {
		jest.clearAllMocks();
		// Mock console.log to prevent test output pollution
		jest.spyOn(console, "log").mockImplementation();
	});

	afterEach(() => {
		// Restore console.log
		jest.restoreAllMocks();
	});

	describe("snapshot tests", () => {
		it("should render correctly with default props", () => {
			const { toJSON } = render(
				<BasicView nameView="TestView">
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			// Verify the component renders correctly
			expect(toJSON()).toMatchSnapshot();
		});

		it("should render correctly with light status bar", () => {
			const { toJSON } = render(
				<BasicView nameView="TestView" statusBarStyle="light">
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			expect(toJSON()).toMatchSnapshot();
		});

		it("should render correctly in full screen mode", () => {
			const { toJSON } = render(
				<BasicView nameView="TestView" isFullScreen={true}>
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			expect(toJSON()).toMatchSnapshot();
		});

		it("should render correctly with all custom props", () => {
			const { toJSON } = render(
				<BasicView
					nameView="TestView"
					statusBarStyle="light"
					isFullScreen={true}
				>
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			expect(toJSON()).toMatchSnapshot();
		});
	});
	describe("functional tests", () => {
		it("should render children correctly", () => {
			// Render with children
			const { toJSON } = render(
				<BasicView nameView="TestView">
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			// Verify children are rendered
			const componentTree = toJSON();
			expect(componentTree).toBeTruthy();
		});

		it("should log view name for re-render analysis", () => {
			const consoleSpy = jest.spyOn(console, "log").mockImplementation();

			render(
				<BasicView nameView="TestView">
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			// Verify console.log was called with the view name
			expect(consoleSpy).toHaveBeenCalledWith("----------TestView---------");
		});

		it("should use correct default values for optional props", () => {
			// Mock StatusBar component
			const mockStatusBar = StatusBar as jest.MockedFunction<typeof StatusBar>;

			// Render with only required props
			render(
				<BasicView nameView="TestView">
					<View>
						<Text>Test Content</Text>
					</View>
				</BasicView>,
			);

			// Verify default values are used
			expect(mockStatusBar).toHaveBeenCalledWith(
				{ style: "dark" }, // Default statusBarStyle
				undefined,
			);
		});

		it("should render ScrollView children correctly", () => {
			const { toJSON } = render(
				<BasicView nameView="TestView">
					<ScrollView>
						<View>
							<Text>Scrollable Content</Text>
						</View>
					</ScrollView>
				</BasicView>,
			);

			// Verify ScrollView is rendered
			const componentTree = toJSON();
			expect(componentTree).toBeTruthy();
		});
	});
});
