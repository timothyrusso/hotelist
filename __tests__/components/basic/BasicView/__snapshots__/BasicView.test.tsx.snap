// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicView snapshot tests should render correctly in full screen mode 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 0,
      "width": "100%",
    }
  }
>
  <View>
    <Text>
      Test Content
    </Text>
  </View>
</View>
`;

exports[`BasicView snapshot tests should render correctly with all custom props 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 0,
      "width": "100%",
    }
  }
>
  <View>
    <Text>
      Test Content
    </Text>
  </View>
</View>
`;

exports[`BasicView snapshot tests should render correctly with default props 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 0,
      "width": "100%",
    }
  }
>
  <View>
    <Text>
      Test Content
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`BasicView snapshot tests should render correctly with light status bar 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 0,
      "width": "100%",
    }
  }
>
  <View>
    <Text>
      Test Content
    </Text>
  </View>
</RCTSafeAreaView>
`;
